### PR TITLE
Create number container system algorithm

### PR DESCRIPTION
**Pull request by CaedenPH**
### Describe your change:
Implements a number container system algorithm which stores indexes and numbers at these corresponding indexes.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: other/number_container_system.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+"""
+A number container system that uses binary search to
+delete and insert values into arrays with O(n logn)
+write times and O(1) read times.
```
Comment (cclauss) [inline other/number_container_system.py:4]: Let' use our 88 character limit...
```suggestion
A number container system that uses binary search to delete and insert values into
arrays with O(n logn) write times and O(1) read times.
```

File: other/number_container_system.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+Further explained in this leetcode problem
+> https://leetcode.com/problems/minimum-cost-tree-from-leaf-values
+"""
+
+
+class NumberContainer:
+    def __init__(self) -> None:
+        # Holds number as the key and returns list of indexes where the number is
+        # The list of indexes is a sorted array in ascending order
```
Comment (cclauss) [inline other/number_container_system.py:16]:
```suggestion
        # numbermap keys are the number and its values are lists of indexes sorted
        # in ascending order
```

File: other/number_container_system.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        self.numbermap: dict[int, list[int]] = {}
+        # Simply holds each index and it's number
```
Comment (cclauss) [inline other/number_container_system.py:18]:
```suggestion
        # indexmap keys are index and it's values are the number at that index
```

File: other/number_container_system.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        self.indexmap: dict[int, int] = {}
+
+    def binary_search_delete(self, array: list[int], item: int) -> list[int]:
+        """
+        Removes the item from the array and returns
+        the new array.
+
+        >>> NumberContainer().binary_search_delete([1,2,3], 2)
+        [1, 3]
+        >>> NumberContainer().binary_search_delete([], 0)
```
Comment (cclauss) [inline other/number_container_system.py:28]: Please add tests for
```
        >>> NumberContainer().binary_search_delete([0, 0, 0], 0)
        >>> NumberContainer().binary_search_delete([-1, -1, -1], -1)
        >>> NumberContainer().binary_search_delete([-1, 0], 0)
        >>> NumberContainer().binary_search_delete([0, -1], 0)
        >>> NumberContainer().binary_search_delete([-1, 0], -1)
        >>> NumberContainer().binary_search_delete([0, -1], -1)
        >>> NumberContainer().binary_search_delete(range(7), 3)
        >>> NumberContainer().binary_search_delete([1.1, 2.2, 3.3], 2.2)
        >>> NumberContainer().binary_search_delete("abcde", "c")
```

**Review by cclauss — CHANGES_REQUESTED** [review]
This looks cool!!

How does it perform on long lists of numbers?
```
from random import shuffle
numbers = list(range(10_000))
shuffle(numbers)
binary_search_delete(numbers, 5_000)
```

**Comment by CaedenPH** [comment]
> This looks cool!!
> 
> 
> 
> How does it perform on long lists of numbers?
> 
> ```
> 
> from random import shuffle
> 
> numbers = list(range(10_000))
> 
> shuffle(numbers)
> 
> binary_search_delete(numbers, 5_000)
> 
> ```

Good idea, I can add some timeit methods to show the scalability of this algorithm.

Note that binary search requires a sorted array so shuffle cannot be used

**Comment by CaedenPH** [comment]
@cclauss I have added more tests to both insert and delete functions along with allowing not only list of ints but also strings and range objects.
I'm not sure if implementing benchmarks is necessary because none of the search or sort algorithms have benchmarks unless they are being compared to other methods

**Review by cclauss — APPROVED** [review]

**Pull request merged**